Intro to git: Content (#40)
Ada-Developers-Academy/core-unit-1: intro-to-git/git-setup.md

## Forking an Existing Git Project For Yourself

The community model for GitHub encourages forking repos.

When developers see a repo that would be a good _base_ for a project, and the developer wants to remix this project for themselves, they can **fork** a project.

Forking creates a copy of the repo (including its code _and_ its Git history up to that point). This copy of the repo will belong to your own user account.

To fork a repo, use GitHub.com:

- Navigate to the repo's page on Github.com
- Locate the button to fork this repo. This button is likely towards the top right corner.
- Confirm that there is a new repo with the same name, but owned by your account
  - You can check the page that lists your repositories, too

### When Do You Fork?

A developer forks a repo when they need these things:

1. A repo with a copy of this Git history so far (and the code with it)
1. More ownership, control, and access over the repo, such as permissions to add collaborators or push commits
1. Separation from the original repo

## Cloning a Repo Onto Your Computer

When developers see a repo on GitHub.com, and want a copy of this repo on their own computer, they can **clone** the repo.

> Cloning will create a new folder onto the computer. This folder will have the name of the repo. This folder will contain all of the folders, files, and Git history that was in the repo.

Author: I'm gonna give some push back on this:

I'd argue that this paragraph has three distinct thoughts that the student absolutely needs to get pretty early on in Git, and can also get quick feedback on. I'm a believer that the longer sentence in this situation can lead to glazing over the details.

```suggestion
Cloning creates a new folder on the local computer. This folder has the same name as the remote repo. Also, this folder contains all of the remote repo's current folders, files, and Git history.
```

?

> Cloning will create a new folder onto the computer. This folder will have the name of the repo. This folder will contain all of the folders, files, and Git history that was in the repo.

Reviewer: Seems reasonable to me!

The process of cloning requires a clone URL that GitHub usually provides. Once you have this clone URL, the command to clone is:

```bash
$ git clone <clone url>
```

To clone a repo, the full process is:

1. Go to the repo's page on GitHub.com

> 1. Locate the button that opens details on downloading the code or cloning the project. This button is likely labeled "Code" and located towards the top right corner.

Author: Hmmmm

I added the "likely" for a "cover your ass" kinda thing in case GitHub changes that verbiage and we forget to maintain this.

Gonna think more and weigh "how likely is it that that would actually happen" vs. "if it happens and we miss updating it, the pain and confusion would be soooooo annoying lol"

(in my experience, directions that require navigating a site... when they update, and our directions aren't updated, it turns students into thinking that _**they're**_ doing something wrong, when really it was our fault for not updating it lol)

> 1. Locate the button that opens details on downloading the code or cloning the project. This button is likely labeled "Code" and located towards the top right corner.

Reviewer: Fair point.  Arguably, when there are instructions for students, I'd expect a staff member to review them for changes before the unit comes up to make sure they're still valid.  Any time external references are made, that becomes a dependency we have to take ownership of.

1. This button opens details about cloning. Select "HTTPS" as a setting for cloning the project.
1. Locate the clone URL inside of this pop-up. Copy this URL.
1. Switch back to your terminal
1. Run the `$ git clone <clone URL>` command
1. Confirm there's a new project folder and `cd` into it

### Clone with HTTPS

In the process above, it specified choosing "HTTPS" as a setting for cloning the project. This setting determines the protocol used for Git commands and operations; "HTTPS" will be more appropriate for our projects.

Callout: No really, Clone with HTTPS
Cloning with SSH is welcome, and is a great and wonderful way to clone! However, it requires additional setup not covered in this curriculum. This curriculum officially recommends cloning with HTTPS for now, but follow your curiosity!

### When Do You Clone?

A developer clones a repo when they need these things:

1. A repo's code and Git history to be downloaded onto their computer

## Configuring Defaults

> ### Configuring the Default Text Editor

Author: yeah!!!!!
